File: BinarySearch/cpp/lower_and_upper_bound.cpp
//
// Created by santosh on 12/21/25.
//

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int lower_bound(vector<int>& nums, int target, int low, int high)
    {
        if (low > high)
        {
            return low;
        }

        int mid = low + (high - low) / 2;

        if (nums[mid] >= target)
        {
            return lower_bound(nums, target, low, mid - 1);
        }
        return lower_bound(nums, target, mid + 1, high);
    }

    int upper_bound(vector<int>& nums, int target, int low, int high)
    {
        if (low > high)
        {
            return low;
        }
        int mid = low + (high - low) / 2;
        if (nums[mid] > target)
        {
            return upper_bound(nums, target, low, mid - 1);
        }
        return upper_bound(nums, target, mid + 1, high);
    }

    int searchInsert(vector<int>& nums, int target,int low, int high)
    {
        if (low > high)
        {
            return low;
        }

        int mid= low + (high - low) / 2;

        if (nums[mid]>=target)
        {
            return lower_bound(nums, target, low, mid - 1);
        }
        return lower_bound(nums, target, mid + 1, high);

    }

    int floorValue(vector<int>& nums, int target, int low, int high)
    {
       return  nums[this->upper_bound(nums, target, low, high)-1];
    }

    int ceilValue(vector<int>& nums, int target, int low, int high)
    {
        return  nums[this->lower_bound(nums, target, low, high)];
    }

    int lastOccurenceInSortedArray(vector<int>& nums, int target, int low, int high)
    {
        int lastOccurenceIndex = this->lower_bound(nums, target, low, high);
        if (nums[lastOccurenceIndex] >= target)
        {
            return lastOccurenceIndex;
        }
        else return -1;

    }



};


int main()
{
    vector<int> nums = {3, 4, 13, 13, 13, 20, 40};
    int target = 60;
    int low =0;
    int high = nums.size()-1;
    Solution s;
    // cout << "Lower bound is "<<  s.lower_bound(nums,target,low,high)<<endl;
    // cout << "Upper bound is " << s.upper_bound(nums,target,low,high)<<endl;
    // cout<< "Search insert is " << s.searchInsert(nums, 5, low,high ) << endl;
    // cout << "ceil value is "<< s.ceilValue(nums, target, low, high) << endl;
    // cout << "floor Value is "<< s.floorValue(nums, target, low, high) << endl;
    cout << "Last occurence of index " <<s.lastOccurenceInSortedArray(nums, target, low, high) << endl;

}

File: BinarySearch/python/lower_and_upper_bound.py
# Lower and Upper Bound + related utilities
# Created by santosh

import bisect


class Solution:
    def lower_bound(self, nums: list[int], target: int, low: int, high: int) -> int:
        """
        Returns the index of the first element >= target.
        Recursive approach. Time: O(log n)
        """
        if low > high:
            return low
        mid = low + (high - low) // 2
        if nums[mid] >= target:
            return self.lower_bound(nums, target, low, mid - 1)
        return self.lower_bound(nums, target, mid + 1, high)

    def upper_bound(self, nums: list[int], target: int, low: int, high: int) -> int:
        """
        Returns the index of the first element > target.
        Recursive approach. Time: O(log n)
        """
        if low > high:
            return low
        mid = low + (high - low) // 2
        if nums[mid] > target:
            return self.upper_bound(nums, target, low, mid - 1)
        return self.upper_bound(nums, target, mid + 1, high)

    def search_insert(self, nums: list[int], target: int, low: int, high: int) -> int:
        """Returns the position where target should be inserted."""
        if low > high:
            return low
        mid = low + (high - low) // 2
        if nums[mid] >= target:
            return self.lower_bound(nums, target, low, mid - 1)
        return self.lower_bound(nums, target, mid + 1, high)

    def floor_value(self, nums: list[int], target: int, low: int, high: int) -> int:
        """Returns the largest element <= target."""
        return nums[self.upper_bound(nums, target, low, high) - 1]

    def ceil_value(self, nums: list[int], target: int, low: int, high: int) -> int:
        """Returns the smallest element >= target."""
        return nums[self.lower_bound(nums, target, low, high)]

    def last_occurrence_in_sorted_array(
        self, nums: list[int], target: int, low: int, high: int
    ) -> int:
        """Returns index of first occurrence of target, or -1 if not found."""
        idx = self.lower_bound(nums, target, low, high)
        if idx < len(nums) and nums[idx] >= target:
            return idx
        return -1


if __name__ == "__main__":
    nums = [3, 4, 13, 13, 13, 20, 40]
    target = 60
    low = 0
    high = len(nums) - 1
    s = Solution()

    # print("Lower bound:", s.lower_bound(nums, target, low, high))
    # print("Upper bound:", s.upper_bound(nums, target, low, high))
    # print("Search insert:", s.search_insert(nums, 5, low, high))
    # print("Ceil value:", s.ceil_value(nums, target, low, high))
    # print("Floor value:", s.floor_value(nums, target, low, high))
    print("Last occurrence index:", s.last_occurrence_in_sorted_array(nums, target, low, high))
